File: lincoln/ruby/lib/core/log.rb
class Log < ObjectTable
	
	@@PRIMARY_KEY = Key.new()
	@@TABLENAME = TableName.new(GLOBALSCOPE, "log")
	
	class Type
	  INFO = 0 
	  WARNING = 1 
	  ERROR = 2
  end
	
	class Field
	  TYPE = 0
	  MESSAGE = 1
  end
  
	@@SCHEMA = [String,String]
    # Enum.class,    // Message type
    # String.class   // Log message

	def initialize(context, stream)
		super(context, @@TABLENAME, @@PRIMARY_KEY, TypeList.new(@@SCHEMA))
		@stream = stream
	end
	
  def insert(tuple)
		log = "LOGTYPE [" + tuple.value(Field::TYPE) + "], "
		log += tuple.value(Field::MESSAGE) + "\n"
		
		@stream << log
		return super.insert(tuple)
	end
end

File: lincoln/java/src/jol/core/Log.java
package p2.core;

import java.io.FilterOutputStream;
import java.io.IOException;

import p2.types.basic.Tuple;
import p2.types.basic.TypeList;
import p2.types.exception.UpdateException;
import p2.types.table.Key;
import p2.types.table.ObjectTable;
import p2.types.table.Table;
import p2.types.table.TableName;

public class Log extends ObjectTable {
	
	public static final TableName TABLENAME = new TableName(GLOBALSCOPE, "log");
	public static final Key PRIMARY_KEY = new Key();
	
	public enum Type{INFO, WARNING, ERROR};
	
	public enum Field{TYPE, MESSAGE};
	public static final Class[] SCHEMA = {
		Enum.class,    // Message type
		String.class   // Log message
	};

	private FilterOutputStream stream;
	
	public Log(Runtime context, FilterOutputStream stream) {
		super(context, TABLENAME, PRIMARY_KEY, new TypeList(SCHEMA));
		this.stream = stream;
	}
	
	@Override
	public boolean insert(Tuple tuple) throws UpdateException {
		String log = "LOGTYPE [" + tuple.value(Field.TYPE.ordinal()) + "], ";
		log       += tuple.value(Field.MESSAGE.ordinal()) + "\n";
		
		try {
			stream.write(log.getBytes());
		} catch (IOException e) {
			throw new UpdateException(e.toString());
		}
		return super.insert(tuple);
	}

}
